epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py

```diff
@@ -716,9 +721,13 @@ def write_hardcoded_parameters(self, params):
             res = '// (none)\n'
             return res
         pardef = pardef.replace('std::complex<','cxsmpl<') # custom simplex complex class (with constexpr arithmetics)
-        parset = parset.replace('std::complex<','cxsmpl<') # custom simplex complex class (with constexpr arithmetics)
+        parset = parset.replace('std::complex<','cxsmpl<') # NB this is really needed twice! (if there are two std::complex on the same line)
         parset = parset.replace('sqrt(','constexpr_sqrt(') # constexpr sqrt (based on iterative Newton-Raphson approximation)
-        parset = parset.replace('pow(','constexpr_pow(') # constexpr sqrt (based on iterative Newton-Raphson approximation)
+        parset = parset.replace('pow(','constexpr_pow(') # constexpr pow
+        parset = parset.replace('atan(','constexpr_atan(') # constexpr atan for BSM #627
+        parset = parset.replace('sin(','constexpr_sin(') # constexpr sin for BSM #627
+        parset = parset.replace('cos(','constexpr_cos(') # constexpr cos for BSM #627
+        parset = parset.replace('tan(','constexpr_tan(').replace('aconstexpr_tan(','atan(') # constexpr tan for BSM #627
         parset = parset.replace('(','( ')
         parset = parset.replace(')',' )')
         parset = parset.replace('+',' + ')
```

Comment:
Stating the obvious here, this start to be a smelling code.
I guess that next time we should 
1) move to re replacement with substitution (will be already a great improvment)
2) potentially change the expression upstream where such rule are likely existing --so likely just need to overwrite how such mathematical expression are converted from the token representation to those expression.

But I agree -> for later

Reply:
Thanks Olivier, yes I agree... there are many many many things that should be done in the code upstream, rather than in the plugin. But lets keep it for later.

(It also depends if we want to keep cudacpp as a plugin, or move it to replace the existing export_cpp.py. I imagine I have a slight preference for keeping the plugin: in that case, "cleaning up" would mean hardcoding all of export_cpp in the plugin, and fixing it immediately instead of overloading methods as it is done here? Anyway, for later...)